pkg/queryfrontend/cache/memcached.go

```diff
@@ -0,0 +1,145 @@
+
+package cache
+
+import (
+	"context"
+	"time"
+
+	cortexcache "github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"gopkg.in/yaml.v2"
+
+	"github.com/thanos-io/thanos/pkg/cacheutil"
+)
+
+const (
+	memcachedDefaultTTL = 24 * time.Hour
+)
+
+// MemcachedCache is a memcached-based cache.
+type MemcachedCache struct {
+	logger    log.Logger
+	memcached cacheutil.MemcachedClient
+	validity  time.Duration
+
+	// Metrics.
+	requests prometheus.Counter
+	hits     prometheus.Counter
+}
+
+type MemcachedResponseCacheConfig struct {
+	Memcached cacheutil.MemcachedClientConfig `yaml:",inline"`
+	Validity  time.Duration                   `yaml:"validity"`
+}
+
+// newMemcachedCache makes a new MemcachedCache and returns Cortex ResultsCacheConfig.
+func newMemcachedCache(conf []byte, logger log.Logger, reg prometheus.Registerer) (*queryrange.ResultsCacheConfig, error) {
+	var config MemcachedResponseCacheConfig
+	if err := yaml.UnmarshalStrict(conf, &config); err != nil {
+		return nil, err
+	}
+
+	memcachedClientConfig, err := yaml.Marshal(config.Memcached)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal content of memcached client configuration")
+	}
+
+	memcached, err := cacheutil.NewMemcachedClient(logger, "response-cache", memcachedClientConfig, reg)
+	if err != nil {
+		return nil, errors.Wrap(err, "create memcached client")
+	}
+
+	c := newMemcachedCacheWithClient(memcached, config.Validity, logger, reg)
+
+	return &queryrange.ResultsCacheConfig{
+		CacheConfig: cortexcache.Config{Cache: c},
+	}, nil
+}
+
+func newMemcachedCacheWithClient(
+	memcached cacheutil.MemcachedClient,
+	validity time.Duration,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *MemcachedCache {
+	c := &MemcachedCache{
+		logger:    logger,
+		memcached: memcached,
+		validity:  validity,
+	}
+
+	if c.validity == 0 {
+		level.Info(logger).Log("msg", "memcached cache valid time set to 0, use 24 hours instead")
+		c.validity = memcachedDefaultTTL
+	}
+
+	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_cache_memcached_requests_total",
+		Help: "Total number of items requests to memcached.",
+	})
+
+	c.hits = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_cache_memcached_hits_total",
+		Help: "Total number of items requests to the cache that were a hit.",
+	})
+
+	level.Info(logger).Log("msg", "created memcached cache")
+
+	return c
+}
+
+// Store data identified by keys.
+// The function enqueues the request and returns immediately: the entry will be
+// asynchronously stored in the cache.
+func (c *MemcachedCache) Store(ctx context.Context, keys []string, bufs [][]byte) {
+	var (
+		firstErr error
+		failed   int
+	)
+
+	for i, key := range keys {
```

Comment:
Implicit here is that  `bufs` MUST be at least as long as `keys`, otherwise the program will panic. Do we want to check this early in the function and log an error?

Reply:
Or just document?

Reply:
Or maybe accept as a parameter a map of type `map[string][][]byte` instead?

Reply:
> Or maybe accept as a parameter a map of type map[string][][]byte instead?

I cannot since this is the cache interface from Cortex... And I need to implement it. https://github.com/cortexproject/cortex/blob/master/pkg/chunk/cache/cache.go#L21

> Implicit here is that bufs MUST be at least as long as keys, otherwise the program will panic. Do we want to check this early in the function and log an error?

Umm, I don't know if this case would happen. But I agree with you that it is good to check length first

Reply:
It's a public method on an exposed member, so it could be called by anyone; so I think it would make sense to check it in our func